### DISCUSSION TITLE
optimization of an auxiliary kernel for a neural network

### DISCUSSION THREAD

**Discussion by aiskhak** [idaholab/moose#18892] [Q&A General] [answered]
I developed an auxiliary kernel that predicts turbulent viscosity at each quadrature point. The prediction is performed by a neural network (torchscript model). Everything works fine, but unfortunately, this slows down the code ~50 times. The workflow is organized in the following way:

loop 1 over elements:
-- Momentum Kernel (solves the momentum equation for fluid dynamics);
---- Material Property (sets all material properties, including turbulent viscosity);
       loop 2 over quadrature points:
-------- Auxiliary Kernel (predicts and assigns turbulent viscosity through an auxiliary variable);

I want to optimize my framework. Ideally, I want to predict turbulent viscosity for all quadrature points simultaneously, but it seems impossible since the quadrature points are only accessible within a particular element. I.e., I cannot predict them before loop 1.

The next way to make it faster is at least to predict simultaneously for all quadrature points within one element. I.e., before loop 2 and then just assign the values within this loop. Let's say I have QUAD9 mesh (9 quadratures for each element). This would make my code ~9 times faster by predicting 9 points simultaneously. However, I am not sure what type of object I can develop for this purpose.

I would be glad to hear some suggestions from you.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Isn't loop 1 over quadrature points as well?
What kind of objects do you need to find into the neural network? velocity and pressure variables? some material properties as well or just variables?

Guillaume

**Reply by aiskhak** [reply]
Yes, probably I did not write this correctly.
Material property kernel does the following:

`computeProperties()` function iteratively (over each quadrature point) assigns:

```
for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    _turbulent_viscosity[_qp] = _turb_visc_auxvar[_qp];
```

where `_turb_visc_auxvar[_qp]` comes from the developed auxiliary kernel. This point-wise prediction is very inefficient.

**Reply by aiskhak** [reply]
I might use any type of input feature in my neural network. But currently, I use velocity gradients.

**Reply by GiudGiud** [reply]
you could overwrite the computeProperties routine in your material property to handle all the qps simulatenously. This is not very hard to do

**Reply by aiskhak** [reply]
There are several other properties calculated in `computeProperties` (using different methods)... Not sure what you mean by overwrite.

**Reply by GiudGiud** [reply]
like this is a virtual function and you `override` it with your derived class s implementation

**Reply by GiudGiud** [reply]
note that you are computing this in an auxvariable then moving it to a material. You could just compute this directly in the material.

**Reply by aiskhak** [reply]
Yes, I tried to do this, but something went wrong there... The code behaves very differently comparing to the point-wise prediction. Maybe it is just a bug. Thanks!

**Comment by YaqiWang** [comment]
Parallelization in MOOSE is done through domain decomposition. You can try letting the aux kernel `execute_on` at `nonlinear` to avoid its evaluating during linear solve. This will probably degrade the Newton convergence. Your understanding on how residual is evaluated in MOOSE seems not accurate. You may want to check out `FEProblemBase::computeResidualTags` function. Note that aux kernels and kernels are evaluated in separate element loops.

**Reply by aiskhak** [reply]
I assume that cannot do `execute_on` at `nonlinear` since it will probably affect the accuracy of the solution (I am interested in transient analysis, for steady-state it would be ok). 
Yes, you are right, the auxiliary kernel has its own loop over quadrature points, my bad. My question is still valid - this loop is point-wise and is crazy slow since forward propagation through a deep neural network is expensive.

**Reply by GiudGiud** [reply]
Execute_on nonlinear will not affect the accuracy, only possible the convergence

**Reply by aiskhak** [reply]
ok, thanks, I will check this
